implement websockets tracker for react

ruddell opened the pull request:
- add Tracker frontend
- add `tracker` in Administration reducer for storing user activities
- add Administration reducer test
- add websocket middleware
- add sockjs and webstomp to package.json

The standard place for websockets in a React+Redux project seems to be in a middleware, a developer normally wants access to the redux store when receiving messages.  The tracker page is now working and uses the new `tracker` key in the Administration reducers for keeping track of user activities.

I'm still new to React on web so there are probably things I'm missing (ex: I couldn't figure out how to subscribe to the route changes so I used `window.onhashchange` to send pages to the tracker).

- Please make sure the below checklist is followed for Pull Requests.

- [ ] [Travis tests](https://travis-ci.org/jhipster/generator-jhipster/pull_requests) are green
- [X] Tests are added where necessary
- [x] Documentation is added/updated where necessary
- [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

pascalgrimaud commented:
Nice label @ruddell :)

deepu105 commented:
Nice, I'll take a look and give my thoughts

On Mon, 5 Mar 2018, 7:29 am Pascal Grimaud, <notifications@github.com>
wrote:

> Nice label @ruddell <https://github.com/ruddell> :)
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/jhipster/generator-jhipster/pull/7228#issuecomment-370322369>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABDlF4HF1nzVrE5M1PP_dZvdzRrc6Hymks5tbNtOgaJpZM4Sbuv6>
> .
>

ruddell commented:
@pascalgrimaud I have no idea how that happened haha

I cleaned up the websocket middleware and added a reducer for keeping track of the websocket messages.  I also added tests for the Administration reducers.

deepu105 commented on generators/client/files-react.js:
```diff
@@ -386,6 +393,7 @@ const files = {
                 'spec/app/modules/account/password/password.reducer.spec.ts',
                 'spec/app/modules/account/settings/settings.reducer.spec.ts',
                 'spec/app/modules/administration/user-management/user-management.reducer.spec.ts',
+                { file: 'spec/app/modules/administration/administration.reducer.spec.ts', method: 'processJsx' },
```
:+1:

deepu105 commented on generators/client/templates/react/src/main/webapp/app/config/websocket-middleware.ts.ejs:
```diff
@@ -0,0 +1,108 @@
+import * as SockJS from 'sockjs-client';
+import * as Stomp from 'webstomp-client';
+import { Observable, Observer } from 'rxjs';
```
we dont have RxJS in dependency defined, where do you get it from?

ruddell replied:
It was coming from another dependency I guess.  `yarn why rxjs` says it comes from `lint-staged#listr`.  

I added it to `package.json` conditionally on websockets.

deepu105 replied:
May be we should also ensure we only import what we need from rxjs, live we do for Angular, since rxjs is quite heavy

ruddell replied:
861kb -> 35kb raw size change according to the Webpack visualizer

deepu105 commented on generators/client/templates/react/src/main/webapp/app/config/websocket-middleware.ts.ejs:
```diff
@@ -0,0 +1,108 @@
+    if (!alreadyConnectedOnce) {
+      receive().subscribe(activity => {
+        return store.dispatch({
+          type: ACTION_TYPES.WEBSOCKET_MESSAGE,
```
can we be specific, `WEBSOCKET_ACTIVITY_MESSAGE`

deepu105 commented on generators/client/templates/react/src/main/webapp/app/modules/administration/tracker/tracker.tsx.ejs:
```diff
@@ -0,0 +1,62 @@
+  activities: any[];
+}
+
+export class TrackerPage extends React.Component<ITrackerPageProps> {
```
the component doesnt have state or use lifecycle methods, convert to a arrow method
```
export const TrackerPage = ({ activities }:ITrackerPageProps) => (
   <div>
   ...
   </div>
);

const mapStateToProps = ({ administration }) => ({
  activities: administration.tracker.activities
});

export default connect(mapStateToProps)(TrackerPage);
```

deepu105 commented on generators/client/templates/react/src/main/webapp/app/shared/layout/header/header.tsx.ejs:
```diff
@@ -9,7 +9,7 @@ import {
   FaHome, FaThList, FaUserPlus, FaUser, <%_ if (enableTranslation) { _%>FaFlag,<%_ } _%> FaHeart,
   FaList, FaTasks, FaDashboard, FaBook, FaWrench, FaSignIn, FaSignOut,
-  FaClockO, FaHddO,
+  FaClockO, FaEye, FaHddO,
```
add the icon on new line with a // tslint:disable-next-line comment on top so that it doesnt cause failure when WebSocket is disabled

deepu105 commented on generators/client/templates/react/src/test/javascript/spec/app/modules/administration/administration.reducer.spec.ts.ejs:
```diff
@@ -0,0 +1,213 @@
+
+import administration, { ACTION_TYPES } from 'app/modules/administration/administration.reducer';
+
+describe('User managment reducer tests', () => {
```
user management is on its own reducer so the test for that should be moved to `administration/user-management` folder

deepu105 replied:
Also look at the `generator-jhipster/generators/entity-client/templates/react/src/test/javascript/spec/app/entities/entity-reducer.spc.ts.ejs` to see how the actions can be tested as well

deepu105 commented on generators/client/templates/react/src/test/javascript/spec/app/modules/administration/administration.reducer.spec.ts.ejs:
```diff
@@ -0,0 +1,373 @@
+      store = mockStore({});
+    });
+
+    describe('getSystemProperties', () => {
```
Wrong name in text

deepu105 commented:
Great work, I added 2 more minor comments. Plz take a look

On Wed, 7 Mar 2018, 2:25 am Jon Ruddell, <notifications@github.com> wrote:

> *@ruddell* commented on this pull request.
> ------------------------------
>
> In
> generators/client/templates/react/src/main/webapp/app/config/websocket-middleware.ts.ejs
> <https://github.com/jhipster/generator-jhipster/pull/7228#discussion_r172716252>
> :
>
> > @@ -0,0 +1,108 @@
> +import * as SockJS from 'sockjs-client';
> +import * as Stomp from 'webstomp-client';
> +import { Observable, Observer } from 'rxjs';
>
> It was coming from another dependency I guess. yarn why rxjs says it
> comes from lint-staged#listr.
>
> I added it to package.json conditionally on websockets.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/jhipster/generator-jhipster/pull/7228#discussion_r172716252>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABDlFyGqHEeVfbnZSVZrsrNpVLe5bvDJks5tbzb5gaJpZM4Sbuv6>
> .
>
